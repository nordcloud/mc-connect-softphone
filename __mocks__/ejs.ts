export const renderFile = jest.fn((arg1, arg2, arg3, callback) => {
  callback(undefined, 'dummyRenderFileHTML');
});
