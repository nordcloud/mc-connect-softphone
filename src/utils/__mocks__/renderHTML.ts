export const renderHTML = jest.fn().mockReturnValue('dummyHTML');
