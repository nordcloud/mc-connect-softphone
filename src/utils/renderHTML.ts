import deepmerge from 'deepmerge';
import { Options, renderFile } from 'ejs';
import { StatusCodes } from 'http-status-codes';

import packageJson from '../../package.json';
import { getStageConsts } from './getStageConsts';

const { OAUTH_BASE_URL } = getStageConsts();
const { ASSETS_URL } = process.env;

if (!ASSETS_URL) {
  throw new Error('Missing ASSETS_URL');
}

export async function renderHTML(
  filename: string,
  data: { [name: string]: any } = {},
  options: Options = {}
) {
  const defaultData = {
    appVersion: packageJson.version,
    ASSETS_URL,
    OAUTH_BASE_URL,
  };

  const defaultOptions = {};

  return {
    statusCode: StatusCodes.OK,
    headers: { 'Content-Type': 'text/html' },
    body: await new Promise<string>((resolve, reject) => {
      renderFile(
        `./tmp/${filename}.ejs`,
        deepmerge(defaultData, data),
        deepmerge(defaultOptions, options),
        (err, str) => {
          if (err) reject(err);
          resolve(str);
        }
      );
    }),
  };
}
