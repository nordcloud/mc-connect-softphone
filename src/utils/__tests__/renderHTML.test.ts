import { Options, renderFile } from 'ejs';

import { renderHTML } from '../renderHTML';

const args = ['dummyFileName', undefined, undefined] as unknown as Parameters<
  typeof renderHTML
>;

beforeEach(() => {
  process.env.ASSETS_URL = 'dummyAssetsUrl';
});

describe('renderHTML', () => {
  it('calls renderFile with correct arguments', async () => {
    await renderHTML(...args);
    expect(jest.mocked(renderFile).mock.calls).toMatchSnapshot();
  });

  it('resolves with a correct value when second arg is present', async () => {
    const argsClone = structuredClone(args);
    argsClone[1] = { dummyDataKey: 'dummyDataValue' };
    renderHTML(...argsClone);
    expect(jest.mocked(renderFile).mock.calls).toMatchSnapshot();
  });

  it('resolves with a correct value when third arg is present', () => {
    const argsClone = structuredClone(args);
    argsClone[2] = { dummyOption: 'dummyOptionVaue' } as Options;
    renderHTML(...argsClone);
    expect(jest.mocked(renderFile).mock.calls).toMatchSnapshot();
  });

  it('resolves with a correct value', () => {
    return expect(renderHTML(...args)).resolves.toMatchSnapshot();
  });

  it('rejects with correct value when ASSETS_URL is missing', () => {
    delete process.env.ASSETS_URL;
    return expect(renderHTML(...args)).rejects.toMatchSnapshot();
  });

  it('rejects with a correct value when renderFile rejects', async () => {
    // @ts-ignore
    jest.mocked(renderFile).mockImplementationOnce((a, b, c, callback) => {
      callback('dummyRenderFileError');
    });
    return expect(renderHTML(...args)).rejects.toMatchSnapshot();
  });
});
