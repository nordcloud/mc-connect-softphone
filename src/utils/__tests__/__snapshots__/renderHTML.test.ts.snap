// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderHTML calls renderFile with correct arguments 1`] = `
[
  [
    "./tmp/dummyFileName.ejs",
    {
      "ASSETS_URL": "dummyAssetsUrl",
      "OAUTH_BASE_URL": "https://nordcloud.okta.com/oauth2/v1",
      "appVersion": "0.0.1",
    },
    {},
    [Function],
  ],
]
`;

exports[`renderHTML rejects with a correct value when renderFile rejects 1`] = `"dummyRenderFileError"`;

exports[`renderHTML rejects with correct value when ASSETS_URL is missing 1`] = `[Error: Missing ASSETS_URL]`;

exports[`renderHTML resolves with a correct value 1`] = `
{
  "body": "dummyRenderFileHTML",
  "headers": {
    "Content-Type": "text/html",
  },
  "statusCode": 200,
}
`;

exports[`renderHTML resolves with a correct value when second arg is present 1`] = `
[
  [
    "./tmp/dummyFileName.ejs",
    {
      "ASSETS_URL": "dummyAssetsUrl",
      "OAUTH_BASE_URL": "https://nordcloud.okta.com/oauth2/v1",
      "appVersion": "0.0.1",
      "dummyDataKey": "dummyDataValue",
    },
    {},
    [Function],
  ],
]
`;

exports[`renderHTML resolves with a correct value when third arg is present 1`] = `
[
  [
    "./tmp/dummyFileName.ejs",
    {
      "ASSETS_URL": "dummyAssetsUrl",
      "OAUTH_BASE_URL": "https://nordcloud.okta.com/oauth2/v1",
      "appVersion": "0.0.1",
    },
    {
      "dummyOption": "dummyOptionVaue",
    },
    [Function],
  ],
]
`;
